openedx/core/djangoapps/programs/tasks.py

```diff
@@ -499,71 +507,93 @@ def _retry_with_custom_exception(username, course_run_key, reason, countdown):
     LOGGER.info(f"Running task award_course_certificate for user {user}")
     try:
         course_key = CourseKey.from_string(course_run_key)
-        # Get the cert for the course key and username if it's both passing and available in professional/verified
-        try:
-            certificate = GeneratedCertificate.eligible_certificates.get(
-                user=user.id,
-                course_id=course_key,
-            )
-        except GeneratedCertificate.DoesNotExist:
-            LOGGER.warning(
-                f"Task award_course_certificate was called for user {user} in course run {course_key} but this learner "
-                "has not earned a course certificate in this course run"
-            )
-            return
+    except InvalidKeyError as exc:
+        error_msg = "Failed to determine course key"
+        raise MaxRetriesExceededError(
+            f"Failed to award course certificate for user {username} for course {course_run_key}. Reason: {error_msg}"
+        ) from exc
 
+    # Get the cert for the course key and username if it's both passing and available in professional/verified
+    try:
+        certificate = GeneratedCertificate.eligible_certificates.get(
+            user=user.id,
+            course_id=course_key,
+        )
+    except GeneratedCertificate.DoesNotExist:
+        LOGGER.warning(
+            f"Task award_course_certificate was called for user {user} in course run {course_key} but this learner "
+            "has not earned a course certificate in this course run"
+        )
+        return
+
+    try:
         if (
-            certificate.mode in CourseMode.CERTIFICATE_RELEVANT_MODES
-            or CourseMode.is_eligible_for_certificate(certificate.mode)
+            certificate.mode not in CourseMode.CERTIFICATE_RELEVANT_MODES
+            and not CourseMode.is_eligible_for_certificate(certificate.mode)
         ):
```

Comment:
Converted this to a guard clause with early exit to make the code easier to read, but I would love extra eyes on my conversion of the positive to the negative conditional, because on the one hand I know I'm right and on the other hand that's an easy thing to get wrong.